kvdb-rocksdb: pass ReadOptions to iterators

ordian opened the pull request:
Passing `prefix_same_as_start` has no effect, since we're not using "Prefix Seek" mode, see https://github.com/facebook/rocksdb/blob/9befbe9b406574e8d8b687799701bd8cebf5163b/include/rocksdb/options.h#L1246-L1251 and https://github.com/facebook/rocksdb/wiki/Prefix-Seek-API-Changes.

ordian commented:
> audit code using the prefix search and figure out if it really needs to; if, as I suspect, we're not using this

we are: [1](https://github.com/paritytech/parity-ethereum/blob/f6c3d4c6951ecf2070bdf7e6a5c6a24edb1a43d6/ethcore/blockchain/src/blockchain.rs#L1006-L1015), [2](https://github.com/paritytech/parity-ethereum/blob/f6c3d4c6951ecf2070bdf7e6a5c6a24edb1a43d6/util/journaldb/src/overlayrecentdb.rs#L281-L287).

> In order to use prefix search users must create a prefix extractor (rocksdb::SliceTransform::create_fixed_prefix(…)) when opening the DB and then pass it along when instantiating an iterator, is that correct?

 yes
> The test_prefix_iterator() test in rust-rocksdb uses a different type though: Options (i.e. ffi::rocksdb_options_t) – do both work?

`set_prefix_extractor` is set on `Options` (this is a global option, which affects the db layout, db is partitioned based on prefix),
`set_prefix_same_as_start` is called on `ReadOptions` when creating an Iterator
https://docs.rs/rocksdb/0.13.0/src/rocksdb/db.rs.html#1153

ordian commented:
> (with more docs/examples illustrating usage)

what do you mean, we haven't changed to public API here

dvdplm commented:
> what do you mean, we haven't changed the public API here

True, but the docs for the existing API are lacking and do not mention how to set up the prefix search properly. I think we want an example [here](https://github.com/paritytech/parity-common/pull/277/files#diff-3a3dcd8bf87a5b5c4097e2d12a129e17R80-R81), docs [here](https://github.com/paritytech/parity-common/pull/277/files#diff-3a3dcd8bf87a5b5c4097e2d12a129e17R93) and probably module level docs discussing the quirks of prefix search too?

dvdplm commented:
> > audit code using the prefix search and figure out if it really needs to; if, as I suspect, we're not using this
> 
> we are: [1](https://github.com/paritytech/parity-ethereum/blob/f6c3d4c6951ecf2070bdf7e6a5c6a24edb1a43d6/ethcore/blockchain/src/blockchain.rs#L1006-L1015), [2](https://github.com/paritytech/parity-ethereum/blob/f6c3d4c6951ecf2070bdf7e6a5c6a24edb1a43d6/util/journaldb/src/overlayrecentdb.rs#L281-L287).

I see. So that code needs to be updated after this PR then? And does it even work right now?

ordian commented:
> So that code needs to be updated after this PR then?

no logic has changed

But I just realised that we changed the logic in #257 https://github.com/paritytech/parity-common/commit/8fb8f13c8084ba8770dcfadca71579eeaac05685#diff-4382bfb7e75959b2b8884268540abfa3R597, previously `iter_from_prefix` returned an iterator over all keys >= `prefix`, now it only returns an iterator of keys starting with the given prefix. I thought it was the expected behavior, but not sure now :/

ordian commented:
> But I just realised that we changed the logic in #257

After a quick search, in parity-ethereum we use iter_from_prefix only in [epoch_transitions](https://github.com/paritytech/parity-ethereum/blob/f6c3d4c6951ecf2070bdf7e6a5c6a24edb1a43d6/ethcore/blockchain/src/blockchain.rs#L1006-L1015) and then we filter it to start with prefix [here](https://github.com/paritytech/parity-ethereum/blob/f6c3d4c6951ecf2070bdf7e6a5c6a24edb1a43d6/ethcore/blockchain/src/blockchain.rs#L520-L522). The same thing happens in substrate:
[usage](https://github.com/paritytech/substrate/blob/fc6914cdb8c322cea8a4591bbf9566dd23bfd143/client/src/leaves.rs#L83), [filter](https://github.com/paritytech/substrate/blob/fc6914cdb8c322cea8a4591bbf9566dd23bfd143/client/src/leaves.rs#L84). So even though we've changed the logic in #257, it doesn't affect parity-ethereum and substrate. The question is: do we want the new behavior or the old one?

dvdplm commented on kvdb-rocksdb/src/iter.rs:
```diff
@@ -76,22 +76,27 @@ pub trait IterationHandler {
 {
-	pub fn new(read_lock: RwLockReadGuard<'a, Option<T>>, col: Option<u32>) -> Self {
-		Self { inner: Self::new_inner(read_lock, |db| db.iter(col)) }
+	pub fn new(read_lock: RwLockReadGuard<'a, Option<T>>, col: Option<u32>, read_opts: &ReadOptions) -> Self {
```
docs

dvdplm replied:
`<insert illuminating example here>.` – needs some editing... ;)

…and there are other methods that now take `ReadOptions` that needs some docs?

dvdplm commented on kvdb-rocksdb/src/iter.rs:
```diff
@@ -76,22 +76,27 @@ pub trait IterationHandler {
 
-	pub fn new_from_prefix(read_lock: RwLockReadGuard<'a, Option<T>>, col: Option<u32>, prefix: &[u8]) -> Self {
-		Self { inner: Self::new_inner(read_lock, |db| db.iter_from_prefix(col, prefix)) }
+	pub fn new_from_prefix(
```
docs

dvdplm commented on kvdb-rocksdb/src/lib.rs:
```diff
@@ -588,12 +588,14 @@ impl Database {
 			None
 		};
-		// workaround for https://github.com/facebook/rocksdb/issues/2343
+		// We're not using "Prefix Seek" mode, so the iterator will return
```
I'd duplicate&extend this in the module level docs.

dvdplm commented on kvdb-rocksdb/src/iter.rs:
```diff
@@ -76,22 +80,31 @@ pub trait IterationHandler {
-	pub fn new(read_lock: RwLockReadGuard<'a, Option<T>>, col: Option<u32>) -> Self {
-		Self { inner: Self::new_inner(read_lock, |db| db.iter(col)) }
+	/// Maps `RwLock<RocksDB>` to `RwLock<DBIterator>`, where
+	/// `DBIterator` iterates over all keys.
```
```suggestion
	/// `DBIterator` iterates over all keys.
	/// In addition to a read lock and a column index, it takes a ref to `ReadOptions` to allow customization of the iterator behaviour, e.g. <insert illuminating example here>.
```

dvdplm commented:
ping @bkchr

bkchr commented on kvdb-rocksdb/src/iter.rs:
```diff
@@ -78,22 +84,35 @@ pub trait IterationHandler {
 	/// Create an `Iterator` over the default DB column or over a `ColumnFamily` if a column number
 	/// is passed.
-	fn iter(&self, col: u32) -> Self::Iterator;
+	/// In addition to a read lock and a column index, it takes a ref to the same `ReadOptions` we
```
Not sure why it talks about the read lock here? That is not part of the interface?

Same below.
